src/app/project/project-cac/project-cac.component.ts

```diff
@@ -0,0 +1,413 @@
+      if (item.checkbox === true) {
+        itemsToExport.push(item);
+      }
+    });
+    let list = [];
+    itemsToExport.map(item => {
+        list.push(item.email);
+    });
+
+    let filteredArray = list.reduce((unique, item) => {
+      return unique.includes(item) ? unique : [...unique, item];
+    }, []);
+
+    let userData = '';
+    filteredArray.map(email => {
+      userData += email + ';';
+    });
+    let selBox = document.createElement('textarea');
+    selBox.style.position = 'fixed';
+    selBox.style.left = '0';
+    selBox.style.top = '0';
+    selBox.style.opacity = '0';
+    selBox.value = userData;
+    document.body.appendChild(selBox);
+    selBox.focus();
+    selBox.select();
+    document.execCommand('copy');
+    document.body.removeChild(selBox);
+    this.openSnackBar('Emails have been copied to your clipboard.', 'Close');
+  }
+
+  public selectAction(action) {
+    switch (action) {
+      case 'createCAC': {
+        this.createCAC();
+        break;
+      }
+      case 'deleteCAC': {
+        this.deleteCAC();
+        break;
+      }
+      case 'copyEmail':
+        this.copyEmail();
+        break;
+      case 'selectAll':
+        let someSelected = false;
+
+        // Safety check
+        if (!this.tableData) {
+          return;
+        }
+
+        this.tableData.data.map((item) => {
+          if (item.checkbox === true) {
+            someSelected = true;
+          }
+        });
+        this.tableData.data.map((item) => {
+          item.checkbox = !someSelected;
+        });
+
+        this.selectedCount = someSelected ? 0 : this.tableData.data.length;
+        this._changeDetectionRef.detectChanges();
+        break;
+      case 'delete':
+        this.deleteCACMember();
+        break;
+      case 'export':
+        this.exportItems();
+        break;
+    }
+  }
+
+  private deleteCAC() {
+    const self = this;
+    this.dialogService.addDialog(ConfirmComponent,
+      {
+        title: 'Deleting this list will be a permanent action.',
+        okOnly: false,
+        message: 'Are you sure you want to delete the project Community Advisory Committee from this project?'
+      }, {
+        backdropColor: 'rgba(0, 0, 0, 0.5)'
+      })
+      .takeUntil(this.ngUnsubscribe)
+      .subscribe(
+        isConfirmed => {
+          if (isConfirmed) {
+            self.loading = true;
+            self._changeDetectionRef.detectChanges();
+            self.projectService.deleteCAC(self.currentProject._id).toPromise()
+            .then(() => {
+              // Update cac data, it was successful
+              self.currentProject.projectCAC = false;
+              self.onSubmit();
+            })
+            .catch((error) => {
+              self.loading = false;
+              self._changeDetectionRef.detectChanges();
+              alert('Failed to remove CAC:' + error);
+            });
+          }
+        }
+      );
+  }
+
+  private createCAC() {
+    const self = this;
```

Comment:
I am not sure `self` is required in this context.

Reply:
It's used for scoped changes when the code further down needs to update the right "this" in the context when asynchronously the promise resovles.